### PR TITLE
ENH: Expose additional window functions

### PR DESCRIPTION
**Pull request by jordan-heemskerk**
- Rename existing private function _cos_win to public function
  general_cosine
- Implement general_hamming as a special case of general_cosine
- Convert existing Hamming and Hann windows to special cases of
  general_hamming

### CODE REVIEW

File: scipy/signal/windows/windows.py [hunk 1]
```diff
@@ -958,6 +951,98 @@ def barthann(M, sym=True):
...
+
+    The generalized Hamming window is constucted by multiplying a rectangular
+    window by one period of a cosine function [1]_.
+
+    Parameters
+    ----------
+    M : int
+        Number of points in the output window. If zero or less, an empty
+        array is returned.
+    a : float
```
Comment (larsoner) [inline scipy/signal/windows/windows.py]: `alpha` would be more readable than `a`

File: scipy/signal/windows/windows.py [hunk 2]
```diff
@@ -958,6 +951,98 @@ def barthann(M, sym=True):
...
+    Notes
+    -----
+    The generalized Hamming window is defined as
+
+    .. math:: w(n) = \alpha - \left(1 - \alpha\right) \cos\left(\frac{2\pi{n}}{M-1}\right)
+              \qquad 0 \leq n \leq M-1
+
+    Both the common Hamming window and Hann window are special cases of the
+    generalized Hamming window with :math:`\alpha` = 0.54 and :math:`\alpha` =
+    0.5, respectively [2]_.
```
Comment (larsoner) [inline scipy/signal/windows/windows.py]: `See Also` links to the related functions would be good

**Review by larsoner — COMMENTED** [review]
Looks like a reasonable start to me. Still needs unit tests, though. One option is to test against hard-coded values computed elsewhere (e.g. Octave) for short array sizes.

File: scipy/signal/windows/windows.py [hunk 3]
```diff
@@ -958,6 +951,102 @@ def barthann(M, sym=True):
...
+
+    >>> plt.figure()
+    >>> plt.title("Frequency Responses")
+    >>> plt.legend()
+    >>> plt.ylabel("Normalized magnitude [dB]")
+    >>> plt.xlabel("Normalized frequency [cycles per sample]")
+    >>> freq_plot = plt.axes()
+
+    >>> for alpha in [0.75, 0.7, 0.52]:
+    >>>    window = general_hamming(41, alpha)
```
[the remark this answers is not in the document]
Comment (larsoner) [reply scipy/signal/windows/windows.py]: I think usually indented lines are preceded by `...` in place of `>>>`

[the remark this answers is not in the document]
Comment (larsoner) [reply scipy/signal/windows/windows.py]: ... meaning, indent the lines inside the `for` loop

[the remark this answers is not in the document]
Comment (larsoner) [reply scipy/signal/windows/windows.py]:
```
    >>> for alpha in [0.75, 0.7, 0.52]:
    ...    window = general_hamming(41, alpha)
```

[the remark this answers is not in the document]
Comment (jordan-heemskerk) [reply scipy/signal/windows/windows.py]: That did the trick, thanks @larsoner

[the remark this answers is not in the document]
Comment (endolith) [reply scipy/signal/windows/windows.py]: Oops. Yes, do that

**Comment by endolith** [comment]
> I haven't encountered these ones in practice yet. I'm happy to take them on as well, but maybe in a different PR?

Yeah this just reminded me of it.  I'll submit it as another PR.  I just have to finish the documentation. 

(I had initially tried to modify `blackman` to handle both cases, but that wouldn't have been backwards compatible, so it sat around for a long time until I realized I could just make separate functions for them.)

File: scipy/signal/windows/windows.py [hunk 4]
```diff
@@ -958,6 +927,96 @@ def barthann(M, sym=True):
...
+    generalized Hamming window with :math:`\alpha` = 0.54 and :math:`\alpha` =
+    0.5, respectively [2]_.
+
+    See Also
+    --------
+    hamming, hann
+
+    Examples
+    --------
+
```
Comment (endolith) [inline scipy/signal/windows/windows.py]: no need for a blank line here

File: scipy/signal/windows/windows.py [hunk 5]
```diff
@@ -958,6 +927,96 @@ def barthann(M, sym=True):
 
     return _truncate(w, needs_trunc)
 
+def general_hamming(M, alpha, sym=True):
+    r"""Return a generalized Hamming window.
+
+    The generalized Hamming window is constucted by multiplying a rectangular
```
Comment (endolith) [inline scipy/signal/windows/windows.py]: typo in "constructed"

File: scipy/signal/windows/windows.py [hunk 6]
```diff
@@ -958,6 +927,96 @@ def barthann(M, sym=True):
...
+    >>> freq_plot = plt.axes()
+
+    >>> for alpha in [0.75, 0.7, 0.52]:
+    ...     window = general_hamming(41, alpha)
+    ...     spatial_plot.plot(window, label="{:.2f}".format(alpha))
+    ...     A = fft(window, 2048) / (len(window)/2.0)
+    ...     freq = np.linspace(-0.5, 0.5, len(A))
+    ...     response = 20 * np.log10(np.abs(fftshift(A / abs(A).max())))
+    ...     freq_plot.plot(freq, response, label="{:.2f}".format(alpha))
+
```
Comment (endolith) [inline scipy/signal/windows/windows.py]: should just be one blank line before the next heading

File: scipy/signal/windows/windows.py [hunk 7]
```diff
@@ -958,6 +927,96 @@ def barthann(M, sym=True):
...
+    values include 0.75, 0.7 and 0.52 [4]_. As an example, we plot these different
+    windows.
+
+    >>> from scipy.signal.windows import general_hamming
+    >>> from scipy.fftpack import fft, fftshift
+    >>> import matplotlib.pyplot as plt
+
+    >>> plt.figure()
+    >>> plt.title("Generalized Hamming Windows")
+    >>> plt.legend()
```
Comment (endolith) [inline scipy/signal/windows/windows.py]: [The legend is blank](https://5932-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.windows.general_hamming.html), you need to call this after plotting the labeled curves

**Comment by jordan-heemskerk** [comment]
@endolith @larsoner Thanks for the reviews - is there anything preventing this from being ready to merge?

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge.

Let's give @endolith or any others a couple more days to look (again) if they want. @jordan-heemskerk feel free to ping me in a couple of days for merge if nobody else suggests changes.

**Comment by endolith** [comment]
It looks fine to me, except maybe move the legend in the freqrep plot to upper corner? https://5986-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.windows.general_hamming.html

**Comment by jordan-heemskerk** [comment]
Moved the legend to the upper right corner now

**Comment by jordan-heemskerk** [comment]
@larsoner ping :)

**Comment by stefanv** [comment]
This looks like a useful addition, thank you.

**Comment by larsoner** [comment]
Thanks @jordan-heemskerk

**Pull request merged**